### PR TITLE
Analysis results

### PR DESCRIPTION
**Pull request by orbitfold**
Create a class for storing results of analyses.

### CODE REVIEW

File: easyvvuq/analysis/pce_analysis.py [hunk 1]
```diff
@@ -2,15 +2,91 @@
...
+
+        Returns
+        -------
+        list of floats
+            Will return a list with two nans, since this is
+        pandas way for handling missing values it seems.
+        """
+        return [float('nan'), float('nan')]
+
+    def describe(self):
```
Comment (raar1) [inline easyvvuq/analysis/pce_analysis.py:78]: This is a minor remark, but I'm not sure ```describe()``` is a very clear name. Without looking at the code I certainly wouldn't have guessed what it does, or that it returns a dataframe. For sure this method should already have a docstring, but ideally we'd choose a better name. ```get_stats_as_dataframe()``` is too verbose, but would something along those lines work? (Please correct me if ```describe``` is actually a common idiom in data science or something like that...)

File: easyvvuq/analysis/pce_analysis.py [hunk 2]
```diff
@@ -2,15 +2,91 @@
 """
 import logging
 import chaospy as cp
+import pandas as pd
 from easyvvuq import OutputType
 from .base import BaseAnalysisElement
+from .results import AnalysisResults
+from .qmc_analysis import QMCAnalysisResults
 
 __author__ = 'Jalal Lakhlili'
```
Comment (raar1) [inline easyvvuq/analysis/pce_analysis.py:11]: I suspect Jalal is no longer the sole author of this file... ;)

File: easyvvuq/analysis/pce_analysis.py [hunk 3]
```diff
@@ -2,15 +2,91 @@
...
+           Quantity of interest
+        input_ : str
+           Input variable
+
+        Returns
+        -------
+        float
+            First order sobol index.
+        """
+        raw_dict = AnalysisResults._keys_to_tuples(self.raw_data['sobols_first'])
```
Comment (raar1) [inline easyvvuq/analysis/pce_analysis.py:35]: I don't think I would really understand what these two lines are doing if I came across them... Essentially it's sort of building the dict according to a 'template' for (in this case) first sobol indices? Is that the right way to think about it, and if so, could we find a higher level naming so it's obvious what's happening here?

File: easyvvuq/analysis/pce_analysis.py [hunk 4]
```diff
@@ -2,15 +2,91 @@
...
+    def get_sobols_total_conf(self, qoi, input_):
+        """Not implemented for this method.
+
+        Returns
+        -------
+        list of floats
+            Will return a list with two nans, since this is
+        pandas way for handling missing values it seems.
+        """
+        return [float('nan'), float('nan')]
```
Comment (raar1) [inline easyvvuq/analysis/pce_analysis.py:76]: It's annoying that we need to return `NaN`s just to integrate with ```pandas```. If a user calls a method that isn't implemented/applicable the best way to handle it is usually to raise an exception. Returning ```NaN``` is kind of ugly and could create problems down the line if a user is not using ```pandas```. Could we add a kwarg called ```suppress_exceptions``` that by default is ```False```, so if a user calls this normally an exception will be raised as it should. But wherever you are feeding this into a pandas dataframe, that code can use ```suppress_exceptions=True``` and get the NaNs it wants? I'm open to other ideas too.

File: easyvvuq/analysis/qmc_analysis.py [hunk 5]
```diff
@@ -17,8 +18,27 @@
...
+        raw_dict = AnalysisResults._keys_to_tuples(self.raw_data['sobols_first'])
+        return raw_dict[AnalysisResults._to_tuple(qoi)][input_][0]
+
+    def get_sobols_total(self, qoi, input_):
+        raw_dict = AnalysisResults._keys_to_tuples(self.raw_data['sobols_total'])
+        return raw_dict[AnalysisResults._to_tuple(qoi)][input_][0]
+
+    def get_sobols_first_conf(self, qoi, input_):
+        raw_dict = AnalysisResults._keys_to_tuples(self.raw_data['conf_sobols_first'])
+        return [raw_dict[AnalysisResults._to_tuple(qoi)][input_]['low'][0],
```
Comment (raar1) [inline easyvvuq/analysis/qmc_analysis.py:34]: Just following on from a previous comment - I'm imagining having written my own, new Analysis element, and then having to create an ```AnalysisResults``` subclass for it. How easy is that to do? Looking at the ```QMCAnalaysisResults``` I might be scared off. Is there not a higher level way we can represent this stuff?

File: easyvvuq/analysis/results.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+                results = results[qoi_cols]
+        else:
+            if qoi_cols:
+                results = self.samples[qoi_cols].describe()
+            else:
+                results = self.samples.describe()
+        return results
+
+    @staticmethod
+    def _keys_to_tuples(dictionary):
```
Comment (raar1) [inline easyvvuq/analysis/results.py:254]: Sorry for being stupid, but why do the keys need to sometimes become tuples? Can't everything just remain a ```dict``` internally to the results objects?

File: easyvvuq/analysis/sc_analysis.py [hunk 7]
```diff
@@ -32,6 +34,31 @@
...
+        return raw_dict[AnalysisResults._to_tuple(qoi)][input_][0]
+
+    def get_sobols_total(self, qoi, input_):
+        return float('nan')
+
+    def get_sobols_first_conf(self, qoi, input_):
+        return [float('nan'), float('nan')]
+
+    def get_sobols_total_conf(self, qoi, input_):
+        return [float('nan'), float('nan')]
```
Comment (raar1) [inline easyvvuq/analysis/sc_analysis.py:45]: Same comment as earlier concerning the ```suppress_exceptions``` kwarg.

File: easyvvuq/utils/helpers.py [hunk 8]
```diff
@@ -29,3 +29,29 @@ def multi_index_tuple_parser(lst):
             contains_tuples = False
         result.append(name)
     return result, contains_tuples
+
+
+def remove_start_of_file(filename, delimiter):
```
Comment (raar1) [inline easyvvuq/utils/helpers.py:34]: Out of curiosity, what is this being used for? I don't see it being used anywhere else in the PR...

**Review by raar1 — CHANGES_REQUESTED** [review]
Good work standardising the results of analysis elements :) I've left a few comments here and there, most of them out of curiosity (not necessarily demanding changes).

The one thing I'd really like to see, though, would be a standard method of ```AnalysisResults``` that returns a list of which methods are 'implemented' for this particular results object. Furthermore, it would be better if (under normal usage) the non-implemented methods would raise an exception when called, and would dump out this list of what methods _are_ available. I think that would be a lot more user friendly. You can still suppress these exceptions and get NaNs for your dataframe in the relevant code sections. Anyway, I've left some comments in the appropriate places to discuss this.

File: easyvvuq/analysis/pce_analysis.py (hunk 1 above)
Comment (dww100) [reply easyvvuq/analysis/pce_analysis.py:78]: In my mind `describe(df)` in R returns a (typically baffling) custom type containing basic stats about the input dataframe. I think the analogy is actually quite intuitive here but I suspect that you are right that it is unclear for most users.

Comment (raar1) [reply easyvvuq/analysis/pce_analysis.py:78]: Ah I see, so it is meaningful to data scientists, at least. HPC devs etc might not immediately recognise it. I guess we can maybe leave it as ```describe()``` for now though. It's not the _worst_ name...

File: easyvvuq/analysis/pce_analysis.py (hunk 2 above)
Comment (jlakhlili) [reply easyvvuq/analysis/pce_analysis.py:11]: You are right. 
Btw, I think we have to clean all `__author__` metadata in the library.

Comment (raar1) [reply easyvvuq/analysis/pce_analysis.py:11]: Yes I think a lot of stuff is very out of date, also with respect to contributors etc. I did add a CITATION.cff file (see issue #274 ) recently, which I'd like to make more inclusive (lots of people have helped or contributed but are not necessarily on the main publications).

**Review by dww100 — APPROVED** [review]
All good

**Review by raar1 — APPROVED** [review]

**Comment by jlakhlili** [comment]
There is an issue in case of vectors results, `describe` returns just the first value of the moments.
https://github.com/UCL-CCS/EasyVVUQ/blob/dev/easyvvuq/analysis/pce_analysis.py#L78
Here is an example with cooling model:

Describes give this:

```
In [28]: results.describe()
Out[28]: 
                 te
10%    9.500000e+01
90%    9.500000e+01
count  2.400000e+03
mean   9.500000e+01
std    3.608792e-12
var    1.302338e-23
```

The results should be an array.

```
n [27]: results.raw_data['statistical_moments']
Out[27]: 
{'te': {'mean': array([95.        , 90.14480222, 85.62853232, 81.42607272, 77.51425404,
         73.87169698, 70.47867197, 67.31696485, 64.36975852, 61.62152061,
         59.05790055, 56.66563881, 54.43247737, 52.34708481, 50.39898016,
         48.57846789, 46.87657596, 45.28499994, 43.79605167, 42.40261031,
         41.09807879, 39.87634425, 38.7317402 , 37.65901235, 36.65328742,
         35.71004408, 34.82508628, 33.99451843, 33.21472289, 32.48233907,
         31.79424428, 31.14753593, 30.53951502, 29.96767121, 29.42966884,
         28.92333398, 28.44664273, 27.99771014, 27.57478008, 27.17621601,
         26.80049218, 26.44618579, 26.11196953, 25.79660476, 25.49893528,
         25.21788137, 24.95243443, 24.70165205, 24.46465327, 24.24061436,
         24.0287648 , 23.82838365, 23.63879613, 23.45937039, 23.28951465,
         23.12867447, 22.97633017, 22.83199453, 22.69521062, 22.56554974,
         22.44260959, 22.32601246, 22.21540368, 22.11045006, 22.0108385 ,
         21.91627469, 21.82648191, 21.74119985, 21.6601836 , 21.58320266,
         21.51004002, 21.4404913 , 21.37436398, 21.31147665, 21.25165828,
         21.19474764, 21.14059263, 21.08904979, 21.0399837 , 20.99326654,
         20.94877763, 20.90640297, 20.86603485, 20.82757148, 20.79091667,
         20.75597943, 20.72267371, 20.69091813, 20.66063564, 20.63175334,
         20.6042022 , 20.57791684, 20.55283534, 20.52889904, 20.50605233,
         20.4842425 , 20.46341959, 20.44353619, 20.42454736, 20.40641044,
         20.38908494, 20.37253246, 20.35671651, 20.34160244, 20.32715735,
         20.31335   , 20.30015067, 20.28753114, 20.27546457, 20.26392545,
         20.25288953, 20.24233372, 20.23223607, 20.2225757 , 20.21333274,
         20.20448825, 20.19602424, 20.18792355, 20.18016985, 20.17274759,
         20.16564195, 20.1588388 , 20.1523247 , 20.14608682, 20.14011294,
         20.13439141, 20.12891111, 20.12366145, 20.11863232, 20.11381408,
         20.10919753, 20.10477389, 20.10053479, 20.09647222, 20.09257857,
         20.08884653, 20.08526916, 20.08183982, 20.07855214, 20.07540009,
         20.07237785, 20.06947991, 20.06670098, 20.064036  , 20.06148014,
         20.0590288 , 20.05667756, 20.0544222 , 20.05225869, 20.05018318]),

```

@orbitfold Maybe it will be better if I fix this and create PR.

cf. Issue #277

**Pull request merged**